### ISSUE TITLE
Windows build currently appears broken

### ISSUE THREAD

**Issue by ritsu1228** [dmlc/tvm#3426]
When trying to build latest master with VS 2017, I am seeing three errors as listed below (they repeatedly appear during the build process). The first two can be fixed by minor changes to the code, but the last one seems to require toolset upgrade.

```
tvm-master\include\tvm/relay/pass.h(456,14): error C2375:  'tvm::relay::Bind': redefinition; different linkage (compiling source file tvm-master\src\relay\backend\graph_runtime_codegen.cc)
tvm-master\include\tvm/relay/expr_functor.h(249): message :  see declaration of 'tvm::relay::Bind' (compiling source file tvm-master\src\relay\backend\graph_runtime_codegen.cc)
```

It seems that `tvm::relay::Bind` declaration appears in both `include\tvm/relay/pass.h` and `include\tvm/relay/expr_functor.h`, but the latter occurrence does not have `TVM_DLL` qualifier. There seems to be only one implementation of this function in `expr_functor.cc`. Adding the qualifier resolves this error.

```
tvm-master\src\runtime\vm\vm.cc(723,14): error C2065:  'uint': undeclared identifier
```

It appears that VS does not have a builtin definition of `uint`. Changing it to, e.g. `uint32_t` resolves this error. 

```
tvm-master\3rdparty\compiler-rt\builtin_fp16.h(76): error C2059: syntax error: '.' (compiling source file tvm-master\src\relay\pass\quantize.cc)
tvm-master\src\relay\pass\pattern_util.h(210): note: see reference to function template instantiation 'DST_T __truncXfYf2__<float,uint32_t,23,uint16_t,uint16_t,10>(SRC_T)' being compiled
        with
        [
            DST_T=uint16_t,
            SRC_T=float
        ] (compiling source file tvm-master\src\relay\pass\quantize.cc)
tvm-master\src\relay\pass\quantize.cc(177): note: see reference to function template instantiation 'tvm::relay::Constant tvm::relay::MakeConstantScalar<int>(tvm::relay::DataType,T)' being compiled
        with
        [
            T=int
        ]
```

This one seems more complicated. VS 2017 does not seem to support designated initializer. VS 2019 can compile this by turning on `std:c++latest`. However, there are weird runtime error when enabling `std:c++latest` - it appears some packed funcs are misplaced, e.g. calls to `_Array` end up with `_raw_pointer` resulting in runtime error.

Another way to resolve this is to use LLVM. With VS 2019, one can directly open CMake files and add `clang_cl_x64` as the new build config. With VS 2017, one can install the [LLVM extension](https://marketplace.visualstudio.com/items?itemName=LLVMExtensions.llvm-toolchain) and use  `cmake -G "Visual Studio 15 2017 Win64" -T "llvm" ..` to generate solutions that use `clang`; however, in this case, I have to disable LLVM linker.

It would be great if the TVM team could look into the Windows build, check what is going and if there is better solutions (hopefully I am not missing anything). The Windows version does not seem to build properly using the official instructions.

**Comment by Ravenwater** [comment]
+1 on a working/supported Windows build of TVM and VTA

**Comment by tqchen** [comment]
Thanks for reporting this problem. Would be great if you can PR first few of your suggested fixes, we will look into the pattern util later

**Comment by ritsu1228** [comment]
Created PR #3429. Rather than adding `TVM_DLL` to `Bind` in `expr_functor.h`, I think it would be better to remove one of these declarations. However, the two do have different doc strings. It would be great if one of the authors could confirm the desired way of handling it.

In fact, the two declarations of `Bind`s can be left as they are for now - (per my investigation) without changes to the pattern util, we need to use Clang to compile for Windows and Clang only throws a warning for this. However, once VC compiler officially supports designated initializer, we still need to fix the declaration for VC compiler to work.

One other thing I am not able to understand is that why turning on `std:c++latest` in VS 2019 messes things up. As described before, some packed funcs from `tvm.dll` appear shifted (e.g. calls to `_Array` go to `_raw_ptr`; `n = tvm.var('n'); m = tvm.var('m'); A = tvm.placeholder((m, n), name='A')` should reproduce this). I did not spend much time on this as it seems like a compiler/linker issue. I would appreciate if someone could figure out what is going on.

**Comment by Adesun** [comment]
same problem on windows.

**Comment by tqchen** [comment]
fixed by #3458
